modules/storage-page-memory/src/main/java/org/apache/ignite/internal/storage/pagememory/mv/AbstractPageMemoryMvPartitionStorage.java

```diff
@@ -834,23 +958,45 @@ public boolean hasNext() {
             }
 
             while (true) {
-                if (!treeCursor.hasNext()) {
-                    iterationExhausted = true;
-                    return false;
+                if (!closeBusyLock.enterBusy()) {
+                    throwStorageClosedException();
                 }
 
-                VersionChain chain = treeCursor.next();
-                ReadResult result = findLatestRowVersion(chain);
+                try {
+                    if (!treeCursor.hasNext()) {
+                        iterationExhausted = true;
+                        return false;
+                    }
 
-                if (result.isEmpty() && !result.isWriteIntent()) {
-                    continue;
-                }
+                    VersionChain chain = treeCursor.next();
+                    ReadResult result = findLatestRowVersion(chain);
 
-                nextRead = result;
-                currentChain = chain;
+                    if (result.isEmpty() && !result.isWriteIntent()) {
+                        continue;
+                    }
 
-                return true;
+                    nextRead = result;
+                    currentChain = chain;
+
+                    return true;
+                } finally {
+                    closeBusyLock.leaveBusy();
+                }
             }
         }
     }
+
+    /**
+     * Throws an exception that the storage is already closed.
+     */
+    protected void throwStorageClosedException() {
+        throw new StorageClosedException("Storage is already closed");
+    }
+
+    /**
+     * Returns the partition ID.
+     */
+    public int partitionId() {
```

Comment:
Please move this method a bit higher, ideally before any data access methods

Reply:
Fix it